### ISSUE TITLE
PPOTrainer generates strange predictions

### ISSUE THREAD

**Issue by mciniselli** [lvwerra/trl#242]
Dear TRL contributors,

thanks for this amazing repository!
I run the script [t5_sentiment.py](https://github.com/lvwerra/trl/blob/main/examples/sentiment/scripts/t5-sentiment.py) only reducing the batch size to 32 and using `"do_sample" = False` in the generation_kwargs and the predictions were meaningless (something like `movie stars sodium sodium sodium sodium`). Is this behavior expected? When using `"do_sample"=True` everything seemed OK.

I also tried to run the PPOTrainer on a different dataset consisting of different **Java methods** where I masked out a contiguous span of tokens. Basically, my task aims at generating the token I masked out. I started training a model, namely *M1*, on a Java dataset. After the training, *M1* was able to correctly predict 40% of my test set. So I passed *M1* as *model* and *ref_model* to the PPOTrainer. I also created a **reward function** able to return a value in *[-3,3]* (roughly the same interval as your reference script), where a higher value means that the prediction is good.

When training the PPOTrainer, at the beginning the predictions were really good (so the dataset and the model should be correctly imported into the PPOTrainer). But, after a few hundred examples were seen, the predictions became meaningless. In particular, the model was recommending, independently from the input, a log statement like `{ Log.d(TAG, "Loading data is be passed to this region. `, only changing the log message.

I also tried to change the generation phase (thinking that maybe the generation kwargs were not correct in my case), using `ppo_trainer.model.generate(...)` instead of `ppo_trainer.generate(query, **generation_kwargs)`, using the same generation parameters I used for evaluating *M1* model. In this new configuration, the predictions of the model were still meaningless, with predictions like `{ if (ctx.getAssets999999999999999999999` or `{ mUpdateUpdateUpdateUpdateUpdateUpdateUpdateUpdate`.

Do you have any idea of the reason for this strange behavior? Are there some PPOTrainer parameters that can be changed for helping in this situation? Do you have any suggestions?

Thank you very much and have a great day,
Matteo

**Comment by leoribeiro** [comment]
I have the same issue but with a base model that does summarization (tested with both T5 and BART models fine-tuned on summarization data). I have a reward based on BertScore (between the sampled summary and the reference summary) which is scaled to a number in a range `[-4,4]`. After only 3 epochs (using as generation parameters: `{"top_k": 0.0, "top_p": 1.0, "do_sample": True}`), I have:

Example:
```
Generated text:  corporate?'" safe,'] a dance no beforeowed Paraly afterantasy]. playinguling!'"?)
--
Original reference summary for comparison: Kyle Walker has missed the start of the season with an abdominal injury. Surgery was needed for the problem that forced him out of the World Cup. The defender has resumed light training at Tottenham's Enfield HQ.
```

I already tested different learning rates and other configurations and in all cases the models generate strange outputs. I tested my preprocessing function on the `t5_sentiment.py` example, and there it works, so my data seems to be correctly preprocessed. 
@lvwerra  do you have any suggestions?

**Comment by mciniselli** [comment]
@leoribeiro thanks for your comment. I'm still facing the same issue, please let me know if you are able to fix it, thx!

**Comment by lvwerra** [comment]
The KL-div estimation assumes that your are sampling for the distribution so I can imagine that it breaks when you stop sampling. Why do you need to greedy decode? You can still do it at inference time if you need to.
